src/lib/polkadotApi.ts

```diff
@@ -0,0 +1,114 @@
+import { ApiPromise, WsProvider } from "@polkadot/api";
+import { getBlockNumber } from "./connections";
+import { Keyring } from "@polkadot/api";
+import { GENESIS_HASHES } from "./connections";
+import {
+    storeBlockNumber,
+    storeValidAccounts,
+    storeConnected,
+    storeToken,
+    dotApi
+} from "./stores";
+import { options } from "@frequency-chain/api-augment";
+
+import type { DotApi } from "$lib/storeTypes";
+import type { KeyringPair } from "@polkadot/keyring/types";
+import type { web3Enable, web3Accounts } from "@polkadot/extension-dapp";
+import type { InjectedAccountWithMeta } from "@polkadot/extension-inject/types";
+import type { ChainProperties } from "@polkadot/types/interfaces";
+
+interface AccountMap {
+    [key: string]: KeyringPair;
+}
+interface MetaMap {
+    [key: string]: InjectedAccountWithMeta;
+}
+
+export async function getApi(
+    selectedProviderURI: string,
+    selectedProvider: string,
+    thisDotApi: DotApi,
+    wsProvider: WsProvider,
+    thisWeb3Enable: typeof web3Enable,
+) {
+
+    if (!selectedProviderURI) {
+        throw new Error("Empty providerURI");
+    }
+    // Handle disconnects
+    if (selectedProviderURI) {
+        if (thisDotApi?.api) {
+            await thisDotApi.api.disconnect();
+        } else if (wsProvider) {
+            await wsProvider.disconnect();
+        }
+    }
+
+    // Singleton Provider because it starts trying to connect here.
+    // Check that the polkadot extension is installed if connecting to Rococo
+    if (selectedProvider === "Rococo") {
+        const extensions = await thisWeb3Enable('Frequency parachain provider dashboard');
+        if (!extensions || !extensions.length) {
+            alert("Polkadot{.js} extension not found; please install it first.");
+            throw new Error("Polkadot{.js} extension not found; please install it first.");
```

Comment:
Moved this check in here from `getApi()` so that we get an alert and an error if the polkadot.js extension is not enabled (instead of a connection error). This was an easy error condition to happen in testing.

Reply:
This will fail if we add mainnet, or if the custom endpoint is a node on some other relay chain. There's a function called isLocalhost that I think would be a better check, because generally that's when we would be using KeyRing and in every other case use the extension.

Reply:
I moved the polkadot extension check back into `loadAccounts`, and used `isLocalhost`.
I briefly tried to disconnect on error but it doesn't want to work, and the user needs to reload with the browser extension to use the polkadot extension anyway.

Reply:
I've had issues with it too. Maybe later we can do a deep-dive and try to do something better.